src/server/lib/logger.ts

```diff
@@ -1,9 +1,68 @@
+import { AWSError, Kinesis } from 'aws-sdk';
 import { Logger, LogLevel } from '@/server/models/Logger';
 import { createLogger, transports } from 'winston';
+import Transport, { TransportStreamOptions } from 'winston-transport';
 import { formatWithOptions, InspectOptions } from 'util';
+import { awsConfig } from './awsConfig';
+import { getConfiguration } from './getConfiguration';
+
+const {
+  stage,
+  aws: { instanceId, kinesisStreamName },
+} = getConfiguration();
+
+// custom "Winston Transport" to send logs to the AWS kinesis stream
+// see https://github.com/winstonjs/winston-transport#usage
+class KinesisTransport extends Transport {
+  private kinesis: Kinesis;
+
+  constructor(opts?: TransportStreamOptions) {
+    super(opts);
+
+    this.kinesis = new Kinesis(awsConfig);
+  }
+
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  public log(info: any, callback: () => void) {
+    const { level, message } = info;
+
+    setImmediate(() => {
+      this.emit('logger', level);
+      if (kinesisStreamName) {
+        this.kinesis
```

Comment:
what happens if the app logs too much, and the kinesis stream starts blocking or is slow to respond?

In 2018 there was a major incident caused by bad logging behaviour on dotcom frontend, so in the Scala apps we use either AsyncAppender with never block = true
https://github.com/guardian/ophan/blob/main/shared-lib/src/main/scala/ophan/util/ElkLogging.scala#L66

or a logging circuit breaker when this can't be used
https://github.com/guardian/identity/blob/main/identity-api/src/main/scala/com/gu/identity/api/logging/SafeBlockingKinesisAppender.scala#L39

Reply:
After discussion we decided to update the aws configuration options so that the following happens:
- `maxRetries = 0` - Set retries to 0 so that aws doesn't attempt to retry the request
- `httpOptions.connectTimeout = 500` - The maximum time in milliseconds that the connection phase of the request should be allowed to take.
- - `httpOptions.timeout = 250` - Sets the socket to timeout after timeout milliseconds of inactivity on the socket. Defaultsed previously two minutes (120000).

Reply:
We also updated the `trackMetric` method to this too.

This should hopefully prevent any promises from leaking, although we may lose a few logs/metrics if they're long running, but this is unlikely and would only happen under huge loads.